streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/SuppressedImpl.java

```diff
@@ -0,0 +1,225 @@
+
+        @Override
+        public StrictBufferConfig<K, V> unboundedBuffer() {
+            return new StrictBCImpl<>(
+                Long.MAX_VALUE,
+                Long.MAX_VALUE,
+                SHUT_DOWN // doesn't matter, given the bounds
+            );
+        }
+
+        @Override
+        public StrictBufferConfig<K, V> spillToDiskWhenFull() {
+            return new StrictBCImpl<>(maxKeys(), maxBytes(), SPILL_TO_DISK);
+        }
+
+        @Override
+        public StrictBufferConfig<K, V> shutDownWhenFull() {
+            return new StrictBCImpl<>(maxKeys(), maxBytes(), SHUT_DOWN);
+        }
+
+        @Override
+        public EagerBufferConfig<K, V> emitEarlyWhenFull() {
+            return new EagerBCImpl<>(maxKeys(), maxBytes());
+        }
+    }
+
+    public static class IntermediateSuppressionImpl<K, V> implements Suppressed.IntermediateSuppression<K, V> {
+        private BufferConfig<K, V, ? extends BufferConfig> bufferConfig = null;
+        private Duration timeToWaitForMoreEvents = null;
+
+        public IntermediateSuppressionImpl() {}
+
+        private IntermediateSuppressionImpl(final IntermediateSuppressionImpl<K, V> from) {
+            timeToWaitForMoreEvents = from.timeToWaitForMoreEvents;
+            this.bufferConfig = from.bufferConfig;
+        }
+
+        @Override
+        public IntermediateSuppression<K, V> emitAfter(final Duration timeToWaitForMoreEvents) {
+            final IntermediateSuppressionImpl<K, V> result = new IntermediateSuppressionImpl<>(this);
+            result.timeToWaitForMoreEvents = timeToWaitForMoreEvents;
+            return result;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public IntermediateSuppression<K, V> bufferConfig(final BufferConfig<K, V, ? extends BufferConfig> bufferConfig) {
+            final IntermediateSuppressionImpl<K, V> result = new IntermediateSuppressionImpl<>(this);
+            result.bufferConfig = bufferConfig;
+            return result;
+        }
+
+        Duration getTimeToWaitForMoreEvents() {
+            return timeToWaitForMoreEvents;
+        }
+    }
+
+    private SuppressedImpl() {}
+
+    public SuppressedImpl(final StrictBufferConfig<K, V> finalResultsConfig) {
+        this.finalResultsConfig = finalResultsConfig;
+    }
+
+    public SuppressedImpl(final IntermediateSuppression<K, V> intermediateSuppression) {
+        this.intermediateSuppression = (IntermediateSuppressionImpl<K, V>) intermediateSuppression;
+    }
+
+    TimeDefinition<K, V> getTimeDefinition() {
+        return timeDefinition;
+    }
+
+    public boolean isFinalResultsSuppression() {
+        return finalResultsConfig != null;
+    }
+
+    boolean isImmediateEmit() {
+        return intermediateSuppression == null
+            || intermediateSuppression.getTimeToWaitForMoreEvents() != null
+            && intermediateSuppression.getTimeToWaitForMoreEvents().toMillis() <= 0L;
```

Comment:
nit: is a negative value valid? If `Duration` does not check this, we should check and throw?

Reply:
I don't see why I need to worry about that here.

Reply:
I guess `<=` is ok (it was a nit) -- was just wondering if we should use `==` instead.